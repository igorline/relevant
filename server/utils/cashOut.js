import Earnings from '../api/earnings/earnings.model';

export async function logCashOut(user, cashOutAmt, next) {
  try {
    const earnings = await Earnings.find({
      user: user._id,
      status: 'paidout',
      earned: { $gt: 0 }
    });
    const totalPreviousPaidout = earnings.reduce(getTotalEarnings, 0);
    const cashOutLog = new Earnings({
      user: user._id,
      cashOutAttempt: true,
      cashOutAmt,
      prevBalance: user.balance,
      endBalance: user.balance - cashOutAmt,
      totalPreviousPaidout,
      legacyAirdrop: user.legacyAirdrop,
      referralTokens: user.referralTokens,
      airdropTokens: user.airdropTokens,
      status: 'in progress',
      source: 'wallet'
    });
    await cashOutLog.save();
    return cashOutLog;
  } catch (err) {
    return next(new Error('Error logging cashOut', err));
  }
}

export function getTotalEarnings(total, currentVal) {
  total += currentVal.earned;
  return total;
}
