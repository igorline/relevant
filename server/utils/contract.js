import { Contract, ethers } from 'ethers';

const erc20abi = [
  {
    constant: true,
    inputs: [
      {
        name: '_owner',
        type: 'address'
      }
    ],
    name: 'balanceOf',
    outputs: [
      {
        name: 'balance',
        type: 'uint256'
      }
    ],
    payable: false,
    type: 'function'
  }
];

export const getTokenBalance = async ({ address, tokenAddress }) => {
  const provider = ethers.getDefaultProvider('mainnet');
  const contract = new Contract(tokenAddress, erc20abi, provider);
  const balance = await contract.balanceOf(address);
  return balance.toString();
};
