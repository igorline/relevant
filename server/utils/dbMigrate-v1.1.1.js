import CommunityMember from 'server/api/community/community.member.model';
import Community from 'server/api/community/community.model';

async function markDeletedMembers() {
  const deletedCom = await Community.find({ inactive: true }, '_id');
  const ids = deletedCom.map(c => c._id);
  return CommunityMember.updateMany(
    { communityId: { $in: ids } },
    { deletedCommunity: true },
    { new: true }
  );
}

async function updateDB() {
  try {
    await markDeletedMembers();
  } catch (err) {
    console.log(err); // eslint-disable-line
  }
}

updateDB();
