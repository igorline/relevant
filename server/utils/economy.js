import Treasury from '../api/treasury/treasury.model';
import User from '../api/user/user.model';
import Post from '../api/post/post.model';
import Invest from '../api/invest/invest.model';
import apnData from '../pushNotifications';
import Notification from '../api/notification/notification.model';
import Earnings from '../api/earnings/earnings.model';

import { INTERVAL_INFLAITION, INIT_COIN, SHARE_DECAY } from '../config/globalConstants';

async function createCoins() {
  let treasury = await Treasury.findOne({});
  let outstanding = 0;
  let users = await User.find({});
  users.forEach(u => outstanding += u.balance);
  treasury.balance = INIT_COIN - outstanding;
  treasury.totalTokens = INIT_COIN;
  treasury.currentShares = 0;
  treasury.rewardFund = 0;
  treasury.postCount = 0;
  treasury.lastRewardFundUpdate = new Date();
  treasury = await treasury.save();
  console.log(treasury);
  return treasury;
}

// function init() {
//   createCoins()
//   .catch(err => console.log(err));
// }
// init();


async function computePostPayout(posts, treasury) {
  // let posts = await Post.find({ paidOut: false, payoutTime: { $lt: now } });
  let updatedPosts = posts.map(async post => {
    // posts w negative relevance don't get rewars
    if (post.relevance <= 0) {
      return await post.save();
    }
    // linear reward curve
    post.payoutShare = post.relevance / treasury.currentShares;
    post.payout = post.balance + treasury.rewardFund * post.payoutShare;
    post.balance = 0;
    post = await post.save();
    return post;
  });
  updatedPosts = await Promise.all(updatedPosts);
  return updatedPosts;
}


async function allocateRewards() {
  let treasury = await Treasury.findOne({});
  let newTokens = INTERVAL_INFLAITION * treasury.totalTokens;
  treasury.totalTokens += newTokens;
  treasury.rewardFund += newTokens;
  return await treasury.save();
}


async function distributeRewards() {
  let treasury = await Treasury.findOne({});
  let now = new Date();
  let posts = await Post.find({ paidOut: false, payoutTime: { $lte: now } });
  let estimatePosts = await Post.find({ paidOut: false, payoutTime: { $gt: now } });

  // decay curren reward shares
  let decay = (now.getTime() - treasury.lastRewardFundUpdate.getTime()) / SHARE_DECAY;
  // console.log(decay)
  // console.log('start shares ', treasury.currentShares);

  treasury.currentShares *= (1 - Math.min(1, decay));
  treasury.postCount *= (1 - Math.min(1, decay));
  // console.log('end shares ', treasury.currentShares);


  // add post relevance to treasury
  posts.forEach(post => {
    if (post.relevance > 0) {
      treasury.currentShares += post.relevance;
      treasury.postCount += 1;
      console.log('average post shares ', treasury.currentShares / treasury.postCount);
    }
    post.paidOut = true;
  });

  treasury.lastRewardFundUpdate = now;
  treasury = await treasury.save();

  let updatedPosts = await computePostPayout(posts, treasury);
  estimatePosts = await computePostPayout(estimatePosts, treasury);

  return updatedPosts;
}

async function rewardUser(props) {
  let { user, reward, post, treasury, type } = props;

  treasury.rewardFund -= reward;
  if (treasury.rewardFund < 0) throw new Error('Reward fund is empty!');
  treasury = await treasury.save();
  user.balance += reward;
  user = await user.save();

  console.log('Awarded ', user.name, ' ', reward, ' tokens for ', post.title);

  let s = reward === 1 ? '' : 's';
  let action = type === 'vote' ? 'upvoting ' : '';
  let text = `You earned ${reward} coin${s} from ${action}this post`;
  let alertText = `You earned ${reward} coin${s} from ${action}a post`;

  console.log()
  await Earnings.updateRewardsRecord({
    user: user._id,
    post: post._id,
    earned: reward,
  });

  Notification.createNotification({
    post: post._id,
    forUser: user._id,
    type: 'reward',
    coin: reward,
    text
  });

  apnData.sendNotification(user, alertText, {});
  return user;
}

async function distributeUserRewards(posts) {
  let treasury = await Treasury.findOne({});
  let payouts = {};
  let updatedUsers = posts.map(async post => {
    let votes = await Invest.find({ post: post._id });
    // compute total vote shares

    let totalWeights = 0;
    votes.forEach(vote => {
      totalWeights += vote.voteWeight;
      console.log(vote.voteWeight);
    });

    console.log('totalWeights ', totalWeights);

    let author = await User.findOne({ _id: post.user }, 'name balance deviceTokens badge');

    // Current code reards author as the first voter
    // alternately we can use a fixed percentage of reward
    // need to edit the way voteWeight is computed in invest.model createVote()

    // let authorShare = Math.max(AUTHOR_SHARE, author.voteWeight / totalWeights);
    // is it better to use post share or post payout?
    // let curationReward = post.payout * (1 - authorShare);

    let authorShare = 1 / totalWeights;
    let authorPayout = Math.floor(authorShare * post.payout);
    let curationReward = post.payout;

    payouts[author._id] = payouts[author._id] ? payouts[author._id] + authorPayout : authorPayout;

    if (authorPayout > 0) {
      await rewardUser({ user: author, reward: authorPayout, treasury, post });
    }


    //  ---------- Curation rewards ------------

    console.log('curationReward', curationReward);


    let updatedVotes = votes.map(async vote => {
      // don't count downvotes
      if (vote.amount <= 0) return;

      let user = await User.findOne({ _id: vote.investor }, 'name balance deviceTokens badge');

      let curationWeight = vote.voteWeight / totalWeights;
      let curationPayout = Math.floor(curationWeight * curationReward);

      console.log('weight ', curationWeight);
      console.log('payout ', curationPayout);
      if (curationPayout === 0) return;

      payouts[user._id] = payouts[user._id] ? payouts[user._id] + curationPayout : curationPayout;

      await rewardUser({ user, reward: curationPayout, treasury, post, type: 'vote' });
    });
    updatedVotes = await Promise.all(updatedVotes);
    return updatedVotes;
  });

  await Promise.all(updatedUsers);
  console.log('Finished distributing rewards, remaining reward fund: ', treasury.rewardFund);

  return payouts;
}

// exports.pendingUserReward(userId) {
//   let now = new Date();
//   let votes = await Invest.find({ payoutDate: { $lt: now } });
// }
//

let computingRewards = false;
exports.rewards = async () => {
  // safeguard
  if (computingRewards) {
    console.log('Computing rewards in progress');
    return null;
  }
  computingRewards = true;
  try {
    let rewardsAllocation = await allocateRewards();
    let updatedPosts = await distributeRewards();
    let payouts = await distributeUserRewards(updatedPosts);

    let displayPosts = updatedPosts.map(p => ({
      title: p.title,
      payout: p.payout,
      payoutShare: p.payoutShare,
      relevance: p.relevance,
    }));
    console.log('\x1b[32m', 'allocated rewards ', rewardsAllocation);
    console.log('\x1b[32m', 'distributed rewards to posts ', displayPosts);
    console.log('\x1b[32m', payouts);
    console.log('\x1b[0m');
    computingRewards = false;
    return payouts;
  } catch (error) {
    console.log('rewards error', error);
    return null;
  }
};

