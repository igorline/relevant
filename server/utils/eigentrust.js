
function eigenTrust(users, params) {
  const defaults = {
    alpha: 0.95,
    error: 0.0001,
    maxIter: 200,
    initialTrust: [],
  };
  let { alpha, error, maxIter, initalTrust } = { ...defaults, ...params };



}

export default eigenTrust;