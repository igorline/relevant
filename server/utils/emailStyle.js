export const notificationStyle = `
<style>
  .post {
    border: 1px solid lightgrey;
    padding: 4px 16px 4px 16px;
    max-width: 600px;
  }
  .post a {
    text-decoration: none;
  }
  .post .head {
    padding-top: 12px;
  }
  .post .body {
    display: block;
    padding: 12px 0;
    text-decoration: none;
    color: black;
  }
  .footer {
    font-size: 10px;
    color: grey;
  }
</style>`;

export const emailStyle = `
<style>
  p, ul {
    font-size: 14px;
    line-height: 22.5px;
    font-family: Georgia;
    max-width: 667px;
  }
  hr {
    border: .5px solid grey;
    max-width: 664px;
    margin: 30px 0;
  }

  p img {
    max-width: 100%;
    margin: auto;
    text-align: center;
  }
  p img.storeButton {
    margin-top: -10px;
    margin-left: -10px;
  }
  p.buttons {
    margin: 30px 0;
  }
  a {
    color: #3E3EFF;
  }
  a span {
    font-weight: bold;
    font-size: 24px;
    border: 3px solid #3E3EFF;
    padding: 15px 20px;
    display: inline-block;
    margin: 15px 0;
    font-family: Arial;
  }
</style>
`;
