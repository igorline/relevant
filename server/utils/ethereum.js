import request from 'request-promise-any';
import { ethers } from 'ethers';
import { INFURA_NETWORK, NETWORK_NUMBER } from 'app/core/config';

const RelevantToken = require('../../app/contracts/RelevantToken.json');

let decimals;
let instance;
let rpcUrl;
let provider;
let network;
let key;
let tokenAddress;
let web3;
let initialized = false;
let wallet;

const pendingTx = {};

process.on('exit', () => {
  // eslint-disable-next-line
  console.log('Un-executed Pending Transactions: ', pendingTx);
});

export const isInitialized = () => initialized;
export const getWeb3 = () => web3;

// SECURITY - this should never by exposed via any APIs!
export const getInstance = () => instance;

export async function init() {
  try {
    // SECURITY - this env var should never by exposed via any APIs!
    if (process.env.NODE_ENV === 'production') {
      key = process.env.OWNER_KEY;
      if (!key) return false;

      provider = ethers.getDefaultProvider(INFURA_NETWORK);
      tokenAddress = RelevantToken.networks[NETWORK_NUMBER]
        ? RelevantToken.networks[NETWORK_NUMBER].address
        : null;
    } else {
      key = process.env.TEST_KEY;
      if (!key) return false;

      rpcUrl = process.env.TEST_RPC;
      provider = new ethers.providers.JsonRpcProvider(rpcUrl);
      network = await provider.getNetwork();
      tokenAddress = RelevantToken.networks[network.chainId].address;
    }

    wallet = new ethers.Wallet(key, provider);
    instance = new ethers.Contract(tokenAddress, RelevantToken.abi, provider);

    // SECURITY - this should never by exposed via any APIs!
    instance = instance.connect(wallet);

    decimals = await instance.decimals();
    initialized = true;
    return true;
  } catch (err) {
    throw err;
  }
}

export async function getBalance(address) {
  if (!instance) return 0;
  const balance = await instance.balanceOf(address);
  return balance.div((10 ** decimals).toString()).toNumber();
}

export async function getParam(param, opt) {
  if (!initialized) await init();

  let value = await instance[param]();
  if (!opt || !opt.noConvert) value = value.div((10 ** decimals).toString());
  if (!opt || !opt.string) value = parseFloat(value.toString());
  return value;
}

export async function getGasPrice() {
  const gasPrice = await request('https://ethgasstation.info/json/ethgasAPI.json');
  const price = JSON.parse(gasPrice);
  console.log('gas price', price.fast); // eslint-disable-line
  return price.fast;
}

// SECURITY - this function should never by exposed via any APIs!
export async function sendTx({ method, args, cancelPendingTx }) {
  try {
    const gasPrice = await getGasPrice();
    const nonce = await provider.getTransactionCount(wallet.address);
    console.log('current nonce', nonce); // eslint-disable-line
    const optNonce = cancelPendingTx ? { nonce } : {};
    const options = {
      gasPrice: gasPrice * 1e8,
      gasLimit: 6e6,
      ...optNonce
    };
    const tx = await instance[method](...args, options);
    pendingTx[method] = tx;
    const result = await tx.wait();
    delete pendingTx[method];
    console.log('status:', result.status); // eslint-disable-line
    console.log(`gas used by ${method}: ${result.gasUsed}`); // eslint-disable-line
    return result;
  } catch (err) {
    throw err;
  }
}

function toBN(num) {
  return ethers.utils.parseUnits((num / 10 ** decimals).toString(), decimals);
}

export async function mintRewardTokens() {
  if (!instance) await init();
  const lastMint = await instance.roundsSincleLast();
  if (lastMint.toNumber() === 0) return null;
  return sendTx({ method: 'releaseTokens', args: [], overWritePending: true });
}

export async function allocateRewards(_amount) {
  return sendTx({ method: 'allocateRewards', args: [toBN(_amount)] });
}

export async function allocateAirdrops(_amount) {
  return sendTx({ method: 'allocateAirdrops', args: [toBN(_amount)] });
}

export async function getNonce(_account) {
  const nonce = await instance.nonceOf(_account);
  return nonce.toNumber();
}

export async function sign(_account, _amount) {
  const nonce = await getNonce(_account);
  // const amnt = ethers.utils.bigNumberify((_amount * 10 ** 18).toString());
  const amnt = ethers.utils.parseUnits(_amount.toString(), decimals);
  const hash = ethers.utils.solidityKeccak256(
    ['uint256', 'address', 'uint256'],
    [amnt.toString(), _account, nonce]
  );
  const ethSig = await wallet.signMessage(ethers.utils.arrayify(hash));
  return ethSig;
}
