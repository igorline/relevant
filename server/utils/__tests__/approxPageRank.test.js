import Relevance from 'server/api/relevance/relevance.model';
import PostData from 'server/api/post/postData.model';
import Invest from 'server/api/invest/invest.model';

import { sanitize, toObject } from 'server/test/utils';
import { getUsers, getPosts, getCommunities } from 'server/test/seedData';
import computePageRank, { computeApproxPageRank } from 'server/utils/pagerankCompute';

// this will define the database name where the tests are run
process.env.TEST_SUITE = 'pagerankCompute-approx';

jest.mock('server/utils/ethereum');

describe('ethRewards', () => {
  let {
    alice,
    bob,
    carol,
    relevant,
    postI1,
    communityId,
    pagerank,
    link1,
    link2,
    link3
  } = {};

  const next = console.log; // eslint-disable-line

  beforeAll(async () => {
    ({ relevant } = getCommunities());
    ({ alice, bob, carol } = getUsers());
    ({ link1, link2, link3, postI1 } = getPosts());
    communityId = relevant._id;
    const votes = [];

    votes[0] = Invest({
      investor: alice._id,
      post: link1._id,
      amount: 1,
      ownPost: false,
      communityId: relevant._id,
      community: relevant.slug
    });

    votes[1] = Invest({
      investor: bob._id,
      post: link2._id,
      amount: 1,
      ownPost: false,
      communityId: relevant._id,
      community: relevant.slug
    });

    votes[2] = Invest({
      investor: carol._id,
      post: link3._id,
      amount: 1,
      ownPost: false,
      communityId: relevant._id,
      community: relevant.slug
    });

    votes[3] = Invest({
      investor: carol._id,
      author: bob._id,
      post: postI1._id,
      amount: 1,
      ownPost: false,
      communityId: relevant._id,
      community: relevant.slug
    });
    await Promise.all(votes.map(v => v.save()));
    // need to run this to give inital rank to admin
    await computePageRank({ community: relevant.slug, communityId, debug: false });
  });

  beforeEach(() => {
    // global.console = { log: jest.fn() }; // hides logs
  });

  describe('computeApproxPageRank', () => {
    test('should upvote and undo', async () => {
      await testCompute({ amount: 0.5, hasAuthor: true });
    });
  });

  describe('computeApproxPageRank', () => {
    test('should downvote and undo', async () => {
      await testCompute({ amount: -0.5, hasAuthor: true });
    });
  });

  describe('computeApproxPageRank', () => {
    test('should upvote and undo with no author', async () => {
      await testCompute({ amount: -1, hasAuthor: false });
    });
  });

  async function testCompute({ amount, hasAuthor }) {
    const relevantVote = {
      relevanceToAdd: 10,
      community: relevant.slug,
      communityId: relevant._id,
      amount
    };

    postI1.data = await PostData.findOne({ post: postI1._id, communityId });
    const startPagerank = postI1.data.pagerank;

    bob.relevance = await Relevance.findOne({ user: bob._id, communityId, global: true });
    const startAuthorRank = bob.relevance.pagerank;

    let vote = new Invest({
      ...relevantVote,
      investor: alice._id,
      author: hasAuthor ? bob._id : null,
      post: postI1._id,
      ownPost: false
    });
    vote = await vote.save();

    alice.relevance = await Relevance.findOne({
      user: alice._id,
      communityId,
      global: true
    });

    let result = await computeApproxPageRank({
      user: alice,
      author: hasAuthor ? bob : null,
      communityId,
      post: postI1,
      vote
    });

    const { author, post } = result;

    const authorSan = sanitize(toObject(author), 'hashedPassword lastVote salt');
    const postSan = sanitize(toObject(post));
    if (hasAuthor) {
      authorSan.relevance.pagerank = Math.round(authorSan.relevance.pagerank * 100) / 100;
      authorSan.relevance.pagerankRaw =
        Math.round(authorSan.relevance.pagerankRaw * 1000) / 1000;
    }

    postSan.data.pagerank = Math.round(postSan.data.pagerank * 100) / 100;
    postSan.data.pagerankRaw = Math.round(postSan.data.pagerankRaw * 1000) / 1000;
    postSan.data.pagerankRawNeg = Math.round(postSan.data.pagerankRawNeg * 1000) / 1000;

    expect({ post: postSan, author: authorSan }).toMatchSnapshot();

    await vote.remove();
    result = await computeApproxPageRank({
      user: alice,
      author,
      communityId,
      post,
      vote,
      undoInvest: true
    });

    pagerank = result.post.data.pagerank;
    expect(pagerank).toBeCloseTo(startPagerank, 2);
    if (hasAuthor) {
      expect(result.author.relevance.pagerank).toBeCloseTo(startAuthorRank, 2);
    }
  }
});
