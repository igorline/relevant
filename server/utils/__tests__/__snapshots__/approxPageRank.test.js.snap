// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ethRewards computeApproxPageRank should downvote and undo 1`] = `
Object {
  "author": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "airdropTokens": 0,
    "balance": 222,
    "bio": "sup bro!",
    "blocked": Array [],
    "blockedBy": Array [],
    "confirmed": false,
    "createdAt": "[SANITIZED]",
    "desktopSubscriptions": Array [],
    "deviceTokens": Array [],
    "email": "bob@bob.com",
    "ethAddress": Array [
      "0xf17f52151EbEF6C7334FAD080c5704D77216b732",
    ],
    "handle": "bob",
    "hashedPassword": "[SANITIZED]",
    "id": "[SANITIZED]",
    "image": "bob-pic.jpg",
    "investmentCount": 0,
    "lastFeedNotification": 1970-01-01T00:00:00.000Z,
    "lastVote": "[SANITIZED]",
    "legacyAirdrop": 0,
    "legacyTokens": 0,
    "lockedTokens": 0,
    "messages": 0,
    "name": "Bob",
    "notificationSettings": Object {
      "desktop": Object {
        "all": false,
      },
      "email": Object {
        "digest": true,
        "general": false,
        "personal": true,
      },
      "mobile": Object {
        "all": false,
      },
    },
    "onboarding": 0,
    "online": false,
    "password": "test",
    "postCount": 0,
    "profile": Object {
      "name": "Bob",
      "role": "user",
    },
    "referralTokens": 0,
    "relevance": Object {
      "__v": 1,
      "_id": "[SANITIZED]",
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "global": true,
      "invites": 0,
      "pagerank": 57.08,
      "pagerankRaw": 0.073,
      "relevance": 0,
      "relevanceRecord": Array [],
      "topTopics": Array [],
      "updatedAt": "[SANITIZED]",
      "user": "b00000000000000000000000",
    },
    "relevanceRecord": Array [],
    "role": "user",
    "salt": "[SANITIZED]",
    "token": Object {
      "_id": "[SANITIZED]",
      "role": "user",
    },
    "tokenBalance": 0,
    "topTopics": Array [],
    "updatedAt": "[SANITIZED]",
    "votePower": 1,
    "webOnboard": Object {
      "onboarding": false,
    },
  },
  "post": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "body": "awesome post #1!",
    "channel": false,
    "commentCount": 0,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "data": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "balance": 0,
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "downVotes": 0,
      "eligibleForReward": false,
      "expectedPayout": 0,
      "hidden": false,
      "isInFeed": false,
      "latestComment": "[SANITIZED]",
      "pagerank": -20.68,
      "pagerankRaw": 0.034,
      "pagerankRawNeg": 0.059,
      "paidOut": false,
      "parentPost": "000000000000000000000003",
      "payOutShare": 0,
      "payout": 0,
      "payoutTime": "[SANITIZED]",
      "post": "000000000000000000000001",
      "postDate": "[SANITIZED]",
      "rank": 0,
      "relevance": 0,
      "relevanceNeg": 0,
      "repost": false,
      "shares": 0,
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "totalShares": 0,
      "type": "post",
      "upVotes": 0,
      "updatedAt": "[SANITIZED]",
    },
    "downVotes": 0,
    "embeddedUser": Object {
      "_id": "[SANITIZED]",
      "handle": "bob",
      "image": "bob-pic.jpg",
      "name": "Bob",
    },
    "flagged": false,
    "flaggedBy": Array [],
    "hidden": false,
    "id": "[SANITIZED]",
    "image": "link_img1.jpg",
    "links": Array [],
    "mentions": Array [],
    "pagerank": 0,
    "paidOut": false,
    "parentPost": "000000000000000000000003",
    "payoutTime": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "seenInFeedNumber": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "title": "postTitle",
    "twitter": false,
    "twitterScore": 0,
    "type": "post",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
    "url": "https://example.com/testPost1",
    "user": "b00000000000000000000000",
    "version": "metaRework",
  },
}
`;

exports[`ethRewards computeApproxPageRank should upvote and undo 1`] = `
Object {
  "author": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "airdropTokens": 0,
    "balance": 222,
    "bio": "sup bro!",
    "blocked": Array [],
    "blockedBy": Array [],
    "confirmed": false,
    "createdAt": "[SANITIZED]",
    "desktopSubscriptions": Array [],
    "deviceTokens": Array [],
    "email": "bob@bob.com",
    "ethAddress": Array [
      "0xf17f52151EbEF6C7334FAD080c5704D77216b732",
    ],
    "handle": "bob",
    "hashedPassword": "[SANITIZED]",
    "id": "[SANITIZED]",
    "image": "bob-pic.jpg",
    "investmentCount": 0,
    "lastFeedNotification": 1970-01-01T00:00:00.000Z,
    "lastVote": "[SANITIZED]",
    "legacyAirdrop": 0,
    "legacyTokens": 0,
    "lockedTokens": 0,
    "messages": 0,
    "name": "Bob",
    "notificationSettings": Object {
      "desktop": Object {
        "all": false,
      },
      "email": Object {
        "digest": true,
        "general": false,
        "personal": true,
      },
      "mobile": Object {
        "all": false,
      },
    },
    "onboarding": 0,
    "online": false,
    "password": "test",
    "postCount": 0,
    "profile": Object {
      "name": "Bob",
      "role": "user",
    },
    "referralTokens": 0,
    "relevance": Object {
      "__v": 1,
      "_id": "[SANITIZED]",
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "global": true,
      "invites": 0,
      "pagerank": 99,
      "pagerankRaw": 0.169,
      "relevance": 0,
      "relevanceRecord": Array [],
      "topTopics": Array [],
      "updatedAt": "[SANITIZED]",
      "user": "b00000000000000000000000",
    },
    "relevanceRecord": Array [],
    "role": "user",
    "salt": "[SANITIZED]",
    "token": Object {
      "_id": "[SANITIZED]",
      "role": "user",
    },
    "tokenBalance": 0,
    "topTopics": Array [],
    "updatedAt": "[SANITIZED]",
    "votePower": 1,
    "webOnboard": Object {
      "onboarding": false,
    },
  },
  "post": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "body": "awesome post #1!",
    "channel": false,
    "commentCount": 0,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "data": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "balance": 0,
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "downVotes": 0,
      "eligibleForReward": false,
      "expectedPayout": 0,
      "hidden": false,
      "isInFeed": false,
      "latestComment": "[SANITIZED]",
      "pagerank": 72.37,
      "pagerankRaw": 0.085,
      "pagerankRawNeg": 0,
      "paidOut": false,
      "parentPost": "000000000000000000000003",
      "payOutShare": 0,
      "payout": 0,
      "payoutTime": "[SANITIZED]",
      "post": "000000000000000000000001",
      "postDate": "[SANITIZED]",
      "rank": 0,
      "relevance": 0,
      "relevanceNeg": 0,
      "repost": false,
      "shares": 0,
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "totalShares": 0,
      "type": "post",
      "upVotes": 0,
      "updatedAt": "[SANITIZED]",
    },
    "downVotes": 0,
    "embeddedUser": Object {
      "_id": "[SANITIZED]",
      "handle": "bob",
      "image": "bob-pic.jpg",
      "name": "Bob",
    },
    "flagged": false,
    "flaggedBy": Array [],
    "hidden": false,
    "id": "[SANITIZED]",
    "image": "link_img1.jpg",
    "links": Array [],
    "mentions": Array [],
    "pagerank": 0,
    "paidOut": false,
    "parentPost": "000000000000000000000003",
    "payoutTime": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "seenInFeedNumber": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "title": "postTitle",
    "twitter": false,
    "twitterScore": 0,
    "type": "post",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
    "url": "https://example.com/testPost1",
    "user": "b00000000000000000000000",
    "version": "metaRework",
  },
}
`;

exports[`ethRewards computeApproxPageRank should upvote and undo with no author 1`] = `
Object {
  "author": null,
  "post": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "body": "awesome post #1!",
    "channel": false,
    "commentCount": 0,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "data": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "balance": 0,
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "downVotes": 0,
      "eligibleForReward": false,
      "expectedPayout": 0,
      "hidden": false,
      "isInFeed": false,
      "latestComment": "[SANITIZED]",
      "pagerank": -51.3,
      "pagerankRaw": 0.034,
      "pagerankRawNeg": 0.103,
      "paidOut": false,
      "parentPost": "000000000000000000000003",
      "payOutShare": 0,
      "payout": 0,
      "payoutTime": "[SANITIZED]",
      "post": "000000000000000000000001",
      "postDate": "[SANITIZED]",
      "rank": 0,
      "relevance": 0,
      "relevanceNeg": 0,
      "repost": false,
      "shares": 0,
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "totalShares": 0,
      "type": "post",
      "upVotes": 0,
      "updatedAt": "[SANITIZED]",
    },
    "downVotes": 0,
    "embeddedUser": Object {
      "_id": "[SANITIZED]",
      "handle": "bob",
      "image": "bob-pic.jpg",
      "name": "Bob",
    },
    "flagged": false,
    "flaggedBy": Array [],
    "hidden": false,
    "id": "[SANITIZED]",
    "image": "link_img1.jpg",
    "links": Array [],
    "mentions": Array [],
    "pagerank": 0,
    "paidOut": false,
    "parentPost": "000000000000000000000003",
    "payoutTime": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "seenInFeedNumber": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "title": "postTitle",
    "twitter": false,
    "twitterScore": 0,
    "type": "post",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
    "url": "https://example.com/testPost1",
    "user": "b00000000000000000000000",
    "version": "metaRework",
  },
}
`;
